src/frequenz/sdk/model/_manager.py

```diff
@@ -0,0 +1,273 @@
+
+    def _build_path(self, key: str) -> str:
+        """Build the model path for the given key, directly using the path from the model_config.
+
+        Args:
+            key: The key for which to build the model path, must be in the model_config.
+
+        Returns:
+            The full path to the model file associated with the given key.
+
+        Raises:
+            KeyError: If the given key is not found in the model_config.
+        """
+        assert self.model_config, "Model configuration must be set."
+        if key not in self.model_config:
+            raise KeyError(f"Key '{key}' not found in the model configuration.")
+        return self.model_config[key]
+
+
+@dataclass
+class _Model(Generic[T]):
+    """Represent a machine learning model."""
+
+    data: T
+    """The machine learning model."""
+
+    path: str
+    """The path to the model file."""
+
+
+class ModelManager(Generic[T]):
+    """Manage machine learning models.
+
+    The model manager acts as a central hub for retrieving models, reloading
+    models, and monitoring model paths for changes.
+    """
+
+    def __init__(self, model_repository: ModelRepository) -> None:
+        """Initialize the model manager.
+
+        Args:
+            model_repository: a model repository that defines the mapping
+                between model keys and paths.
+        """
+        self._models: dict[Any, _Model[T]] = {}
+
+        models_config = model_repository.get_models_config()
+        for key, model_path in models_config.items():
+            self._models[key] = _Model(data=self._load(model_path), path=model_path)
+
+        self._monitoring_task: asyncio.Task[None] = asyncio.create_task(
+            self._start_model_monitor()
+        )
+
+    def __getitem__(self, key: Any) -> T:
+        """Get a specific loaded model using the subscript operator.
+
+        Args:
+            key: the key to identify the model.
+
+        Returns:
+            the model instance corresponding to the key.
+        """
+        return self.get_model(key)
+
+    def get_model(self, key: Any = None) -> T:
+        """Get a specific loaded model.
+
+        Args:
+            key: the key to identify the model.
+
+        Raises:
+            KeyError: if the key is invalid or it is not provided for
+                multi-model retrieval.
+
+        Returns:
+            the model instance corresponding to the key.
+        """
+        if key is None:
+            if len(self._models) != 1:
+                raise KeyError("No key provided for multi-model retrieval")
+            key = list(self._models.keys())[0]
+
+        if key not in self._models:
+            raise KeyError("Invalid key")
+
+        return self._models[key].data
+
+    def get_paths(self) -> list[Path | str]:
+        """Get the paths of all loaded models.
+
+        Returns:
+            the paths of all loaded models.
+        """
+        return [model.path for model in self._models.values()]
```

Comment:
This returns an iterator instead of a list:
```suggestion
        return (model.path for model in self._models.values())
```